feat: add auto destroy duration docs (#493)
hashicorp/terraform-docs-common: website/docs/cloud-docs/workspaces/settings/deletion.mdx

@include 'tfc-package-callouts/ephemeral-workspaces.mdx'
<!-- END: TFC:only name:pnp-callout -->

> Configuring automatic infrastructure destruction requires admin permissions on at least the workspace.

Reviewer: Adding a link for extra clarity:
```suggestion
Configuring automatic infrastructure destruction for a workspace requires [admin permissions](/terraform/cloud-docs/users-teams-organizations/permissions#workspace-admins) for that workspace.
```

There are two main ways to automatically destroy a workspace's resources:
* Schedule a run to destroy all resources in a workspace at a specific date and time.
* Configure Terraform Cloud to destroy a workspace's infrastructure after a period of workspace inactivity.

You can reduce your spending on infrastructure by automatically destroying temporary resources like development environments.

After Terraform Cloud performs an auto-destroy run, it unsets the `auto-destroy-at` field on the workspace. If you continue using the workspace, you can schedule another future auto-destroy run to remove any new resources. 

!> **Note:** Automatic destroy plans _do not_ prompt you for apply approval in the Terraform Cloud user interface. We recommend only using this setting for development environments.

You can schedule an auto-destroy run using the Terraform Cloud web user interface, or the [workspace API](/terraform/cloud-docs/api-docs/workspaces).

> You can schedule [notifications](/terraform/cloud-docs/workspaces/settings/notifications) to alert you 12 and 24 hours before an auto-destroy run, and to report auto-destroy run results.

Reviewer: On second read, could you move this sentence to the above paragraph? I think the two tie in together sentence tie into each other nicely:
```suggestion
You can also schedule [notifications](/terraform/cloud-docs/workspaces/settings/notifications) to alert you 12 and 24 hours before an auto-destroy run, and to report auto-destroy run results.
```

#### Destroy at a specific day and time

To schedule an auto-destroy run at a specific time in Terraform Cloud:
1. Navigate to the workspace's **Settings** > **Destruction and Deletion** page.
1. Under **Automatically destroy**, click **Set up auto destroy**.
1. Enter the desired date and time. Terraform Cloud defaults to your local time zone for scheduling and displays how long until the scheduled operation.
1. Click **Confirm auto destroy**.

To cancel a scheduled auto-destroy run in Terraform Cloud:
1. Navigate to the workspace's **Settings** > **Destruction and Deletion** page.
1. Under **Automatically destroy**, click **Edit** next to your scheduled run's details.
1. Click **Remove**.

#### Destroy if a workspace is inactive

You can configure Terraform Cloud to automatically destroy a workspace's infrastructure after a period of inactivity.
A workspace is _inactive_ if the workspace's state has not changed within your designated time period.

> !> **Note:** As opposed to configuring an auto-destroy run for a specific date and time, this setting will _persist_ after queued auto-destroy runs.

Reviewer: Love this!
```suggestion
!> **Caution:** As opposed to configuring an auto-destroy run for a specific date and time, this setting _persists_ after queueing auto-destroy runs.
```

> If you configure a workspace to auto-destroy its infrastructure when inactive, any run that updates Terraform state further delays the scheduled auto-destroy time by the length of your designated time period.

Reviewer: Suggesting changing this to "timeframe" because that's what the UI uses:
```suggestion
If you configure a workspace to auto-destroy its infrastructure when inactive, any run that updates Terraform state further delays the scheduled auto-destroy time by the length of your designated timeframe.
```

To schedule an auto-destroy run after a period of workspace inactivity:
1. Navigate to the workspace's **Settings** > **Destruction and Deletion** page.
1. Under **Automatically destroy**, click **Set up auto destroy**.
1. Click the **Destroy if inactive** toggle.

> 1. Enter the desired activity duration.

Reviewer: ```suggestion
1. Select or customize a desired timeframe of inactivity.
```

1. Click **Confirm auto destroy**.

When configured for the first time, the auto-destroy duration setting displays the scheduled date and time that Terraform Cloud will perform the auto-destroy run.
Subsequent auto-destroy runs and Terraform runs that update state both update the next scheduled auto-destroy date.

After Terraform Cloud completes a manual or automatic destroy run, it waits until further state updates to schedule a new auto-destroy run.

> To remove an auto-destroy duration:

Reviewer: ```suggestion
To remove your workspace's auto-destroy based on inactivity:
```

@include 'tfc-package-callouts/ephemeral-workspaces.mdx'
<!-- END: TFC:only name:pnp-callout -->

> 

Reviewer: Do users need specific permissions to perform auto destroy runs? I realize the original page didn't have that info, but I think it'd be good to have here!

Reviewer: Also I'm sorry to be annoying but could you do a find and replace and replace "auto destroy" with "auto-destroy" everywhere on this page? I apologize for making you do this, but I definitely think it's worth it to fix! 🙏

Author: > Do users need specific permissions to perform auto destroy runs? I realize the original page didn't have that info, but I think it'd be good to have here!

Added this, let me know if it should be changed at all
```md
!> **Note:** Automatic destroy plans _do not_ prompt you for apply approval in the Terraform Cloud user interface. We recommend only using this setting for development environments.
```

> Also I'm sorry to be annoying but could you do a find and replace and replace "auto destroy" with "auto-destroy" everywhere on this page? I apologize for making you do this, but I definitely think it's worth it to fix! 🙏

Done! Any issues with either how it's display on the page right now (like button text-wise) or with there being a disconnect between the two?

<img width="887" alt="image" src="https://github.com/hashicorp/terraform-docs-common/assets/33099255/c10cb91c-9b81-4ee0-a763-854c5526b495">

Reviewer: ...well now I want to put in a request to update that UI copy. I think auto-destroy reads more clearly as a compound word. 

This is definitely my fault for not catching this the first time 😅 Do you happen to know who I should talk to or where I should request that change? No worries if not! I can try to figure it out

Author: @Ashtronaut09 any issues with changing all of the UI copy from `auto destroy`->`auto-destroy`? It should be a pretty easy change if everything's all good.

Author: Updating UI copy as part of #17767

There are two main ways of configuring auto destroy for a workspace.
1. Schedule a run to destroy all resources in a workspace at a specific date and time.
1. Allow Terraform Cloud to the destruction of a workspace's infrastructure after a period of workspace inactivity.

> This ephemeral workspace entitlement reduces your infrastructure

Author: Sounds good, I think this line was just moved from above. Updated here because it for sure feels clearer.

You can schedule an auto destroy run using the Terraform Cloud web UI, or the [workspace API](/terraform/cloud-docs/api-docs/workspaces).

You can schedule [notifications](/terraform/cloud-docs/workspaces/settings/notifications) to alert you 12 and 24 hours before an auto destroy run, and to report auto destroy run results.

> #### Destroy at Specific Time

Reviewer: ```suggestion
#### Destroy at a specific day and time
```

1. Under **Automatically destroy**, click **Edit** next to your scheduled run's details.
1. Click **Remove**.

> #### Destroy if Inactive

Reviewer: ```suggestion
#### Destroy if a workspace is inactive
```

> You can configure Terraform Cloud to automatically destroy a workspace's infrastructure after a period of inactivity. Any runs that update Terraform state will push back the scheduled auto destroy time by the configured activity duration designated by the user.

Reviewer: Naive question, but can _other_ workspace affect "inactivity" for a workspace?

Reviewer: I'd love to get super clear on what "inactivity" for a workspace looks like with something like the below suggestion.

```suggestion
You can configure Terraform Cloud to automatically destroy a workspace's infrastructure after a period of inactivity. A workspace is _inactive_ if the workspace's state has not changed within your designated time period.

If you configure a workspace to auto-destroy its infrastructure when inactive, any run that updates Terraform state further delays the scheduled auto-destroy time by the length of your designated time period.
```

Author: >Naive question, but can other workspace affect "inactivity" for a workspace?

Not really, unless there are something like run triggers but those will trigger new runs for the workspace so that _should_ technically be captured here. But open to suggestions if you feel like it should be more explicit.

Reviewer: That makes sense to me! Thank you for explaining the background here, I agree that what you've written covers things!

> 

Reviewer: Could you also add an extra sentence here making it super clear that this setting _persists_ after the first auto-destroy run, unlike the other auto-destroy setting?

Author: Added this, let me know if it needs to any updates

```md
!> **Note:** As opposed to configuring an auto-destroy run for a specific date and time, this setting will _persist_ after queued auto-destroy runs.
```

> To schedule an auto destroy run after a period of inactivity in Terraform Cloud:

Reviewer: ```suggestion
To schedule an auto-destroy run after a period of workspace inactivity:
```

1. Navigate to the workspace's **Settings** > **Destruction and Deletion** page.
1. Under **Automatically destroy**, click **Set up auto destroy**.
1. Click the **Destroy if inactive** toggle.

> 1. Enter the desired activity duration.

Reviewer: Does this populate a suggested default? Or do we have a suggested default? Just curious!

Author: There's a select with the default being `Last 1 day` on creation 
<img width="224" alt="image" src="https://github.com/hashicorp/terraform-docs-common/assets/33099255/a325dfac-729f-4cbc-8617-0a6c9b3e24ba">

with multiple presets to choose from
<img width="215" alt="image" src="https://github.com/hashicorp/terraform-docs-common/assets/33099255/a23cdf5e-8f90-4a61-8b0e-c3b370ef419d">

and a more granular `Custom` option
<img width="411" alt="image" src="https://github.com/hashicorp/terraform-docs-common/assets/33099255/0858848d-52f0-4264-99b0-0f168a8b82ee">

Reviewer: Oooh thank you so much for the _screenshots_! That really helps! Adding suggestions based on this and sorry for not just going and looking at it myself 😅

1. Click **Confirm auto destroy**.

> When configured for the first time, the auto destroy duration setting will display a date and time on which an auto destroy run is scheduled to be performed.

Reviewer: ```suggestion
When configured the first time, the auto destroy duration setting displays the scheduled date and time that Terraform Cloud will perform the auto-destroy run.
```

> This auto destroy date will be updated upon subsequent runs.

Reviewer: ```suggestion
Subsequent auto-destroy runs and Terraform runs that update state both update the next scheduled auto-destroy date.
```

> Each run that updates state will push out the scheduled auto destroy time by the length of time configured in the duration setting.

Reviewer: I think we can remove this sentence because you say this in the above paragraph:
```suggestion
```

> After a manual or automatic destroy run is completed, Terraform Cloud will wait until further state updates to schedule a new auto destroy run.

Reviewer: small rephrase for active voice:
```suggestion
After Terraform Cloud completes a manual or automatic destroy run, it waits until further state updates to schedule a new auto-destroy run.
```

> 

Reviewer: Can we add a section on how to turn this setting off for a workspace?

Reviewer: Perfect, thank you!
